src/SmallStrainPlasticity/SmallStrainPlasticity.jl

```diff
@@ -0,0 +1,332 @@
+\\dot{\\epsilon}_{\\mathrm{p}} = \\dot{\\lambda} \\frac{\\partial \\Phi}{\\boldsymbol{\\sigma}}
+= \\dot{\\lambda} \\boldsymbol{\\nu}
+```
+
+The plastic multiplier, ``\\lambda``, is obtained via the KKT-conditions,
+```math
+\\dot{\\lambda} \\geq 0, \\quad \\Phi \\leq 0, \\quad \\dot{\\lambda}\\Phi = 0
+```
+
+The isotropic hardening is formulated as
+```math
+\\kappa = \\sum_{i=1}^{N_{\\mathrm{iso}}} g_{\\mathrm{iso},i}(\\lambda)
+```
+where ``g_{\\mathrm{iso},i}(\\lambda)`` is specified by `m.isotropic[i]`
+
+Kinematic hardening is formulated as
+```math
+\\boldsymbol{\\beta}_i = \\dot{\\lambda} g_{\\mathrm{kin},i}(\\nu, \\boldsymbol{\\beta}_i)
+```
+where ``g_{\\mathrm{kin},i}(\\boldsymbol{\\nu}, \\boldsymbol{\\beta}_i)`` is specified by `m.kinematic[i]`
+and ``i\\in[1,N_\\mathrm{kin}]``. 
+
+```
+# Keyword arguments
+- `cache`: Cache for the iterative solver, used by NLsolve.jl. It is strongly recommended to pre-allocate the cache for repeated calls to `material_response`. See [`get_cache`](@ref).
+- `options::Dict{Symbol, Any}`: Solver options for the non-linear solver. Under the key `:nlsolve_params` keyword arguments for `nlsolve` can be handed over.
+See [NLsolve documentation](https://github.com/JuliaNLSolvers/NLsolve.jl#common-options). By default the Newton solver will be used.
+"""
+function material_response(material::Chaboche, ϵ::SymmetricTensor{2,3}, state_old::ChabocheState{Nkin,T,N}, Δt; cache=get_cache(material), options::Dict{Symbol, Any} = Dict{Symbol, Any}()) where {T,N,Nkin}
+    
+    σ_trial, dσdϵ_elastic, _, _ = material_response(material.elastic, ϵ-state_old.ϵₚ, nothing, Δt)
+
+    Φ_trial = yield_function(material, σ_trial-sum(state_old.β), state_old.λ)
+    if Φ_trial < 0
+        return σ_trial, dσdϵ_elastic, state_old, true
+    else
+        num_blas_threads = LinearAlgebra.BLAS.get_num_threads() # (~ 2ns)
+        LinearAlgebra.BLAS.set_num_threads(1)                   # Big performance benefit, takes ~8ns
+        converged = solve_local_problem!(cache, material, state_old, ϵ, options)
+        if converged
+            σ, dσdϵ, state = get_plastic_output(cache, material, state_old, σ_trial, dσdϵ_elastic)
+        else
+            σ, dσdϵ, state = (σ_trial, dσdϵ_elastic, state_old)
+            println("Did not converge!")
+        end
+        LinearAlgebra.BLAS.set_num_threads(num_blas_threads)
+        return σ, dσdϵ, state, converged
+    end
+    
+end
+
+# General residual function 
+function residual(X::ChabocheResidual{NKin_R}, material::Chaboche, old::ChabocheState, ϵ) where{NKin_R}
+    Δλ = X.λ - old.λ
+    
+    σ_vm = vonmises_dev(X.σ_red_dev)
+    ν = X.σ_red_dev * ((3/2)*σ_vm)
+
+    Φ = yield_function(material, X.σ_red_dev, X.λ)
+
+    σ_dev = calc_sigma_dev(material.elastic, old, ϵ, ν, Δλ)
+    
+    if NKin_R > 0
+        β_hat0 = σ_dev - X.σ_red_dev - sum(X.β1)
+        β_hat1 = X.β1
+        β1 = ntuple(i->old.β[i+1] + Δλ * get_evolution(material.kinematic[i+1], ν, β_hat1[i]), Val{NKin_R}())
```

Comment:
Now I see the issue: I understand now why this is so hard to read!
NKin_R should probably be NumKinematicInResidual or something, so it is always NKin-1. 
I guess we could dispatch on NKin instead and then use the NKin-1 instead of NKin_R (I removed tha Val-stuff)
The reason for i+1 for old.beta and material.kinematic is that these are tuples of all back-stresses, whereas beta_hat1 is a tuple of only the back-stresses in the residual. I am open for suggestions for how to do this nicer, but the first back-stress (or at least one of them) needs to be treated different from the others. Perhaps the code would look a bit nicer if we use the last instead of the first...
So my notation in this function have been beta*0 indicate the first, but I think this should be beta*1. 
beta*1 indicate the remaining back-stresses...

This is getting so complicated so I think that we should consider using a 13x13 system instead of the 7x7! I'll try to do some benchmarking when I have time.